File: bms/go/geo/geo_s2_Cap_decode/geo_s2_Cap_decode.go
package test

import (
	"encoding/binary"
	"io"
	"math"
)

func (c *geo_s2_Cap) decode(d *geo_s2_decoder) {
	c.center.X = d.readFloat64()
	c.center.Y = d.readFloat64()
	c.center.Z = d.readFloat64()
	c.radius = geo_s1_ChordAngle(d.readFloat64())
}

func (d *geo_s2_decoder) readFloat64() float64 {
	if d.err != nil {
		return 0
	}
	buf := d.buffer()
	_, d.err = io.ReadFull(d.r, buf)
	return math.Float64frombits(binary.LittleEndian.Uint64(buf))
}

func (d *geo_s2_decoder) buffer() []byte {
	if d.buf == nil {
		d.buf = make([]byte, 8)
	}
	return d.buf
}

type geo_s2_Cap struct {
	center geo_s2_Point
	radius geo_s1_ChordAngle
}

type geo_s2_Point struct {
	geo_r3_Vector
}

type geo_r3_Vector struct {
	X, Y, Z float64
}

type geo_s1_ChordAngle float64

type geo_s2_decoder struct {
	r   geo_s2_byteReader // the real reader passed to Decode
	err error
	buf []byte
}

type geo_s2_byteReader interface {
	io.Reader
	io.ByteReader
}

File: translations_only/fail/go/geo/geo_s2_Cap_decode/gemini/CAPR/geo_s2_Cap_decode.rs

use std::io::{Read, Seek, SeekFrom};

pub struct GeoS2Cap {
    pub center: GeoS2Point,
    pub radius: GeoS1ChordAngle,
}

pub struct GeoS2Point {
    pub x: f64,
    pub y: f64,
    pub z: f64,
}

pub struct GeoS1ChordAngle(pub f64);

pub struct GeoS2Decoder<R: Read + Seek> {
    r: R,
    err: Option<std::io::Error>,
    buf: [u8; 8],
}

impl<R: Read + Seek> GeoS2Decoder<R> {
    pub fn new(r: R) -> Self {
        Self { r, err: None, buf: [0; 8] }
    }

    pub fn read_float64(&mut self) -> f64 {
        if let Some(err) = &self.err {
            return 0.0;
        }

        self.r.read_exact(&mut self.buf).unwrap();
        f64::from_le_bytes(self.buf)
    }

    pub fn buffer(&mut self) -> &mut [u8] {
        &mut self.buf
    }
}

impl GeoS2Cap {
    pub fn decode<R: Read + Seek>(&mut self, d: &mut GeoS2Decoder<R>) {
        self.center.x = d.read_float64();
        self.center.y = d.read_float64();
        self.center.z = d.read_float64();
        self.radius = GeoS1ChordAngle(d.read_float64());
    }
}
